src/udev/net/link-config.c

```diff
@@ -1314,6 +1362,58 @@ int config_parse_wol_password(
         return 0;
 }
 
+int config_parse_rps_cpu_mask(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        LinkConfig *config = ASSERT_PTR(userdata);
+        CPUSet *mask = NULL;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+
+        if (isempty(rvalue)) {
+                config->rps_cpu_mask = cpu_set_free(config->rps_cpu_mask);
+                return 0;
+        }
+
+        if (!config->rps_cpu_mask) {
+                mask = new0(CPUSet, 1);
+                if (!mask)
+                        return log_oom();
+
+                config->rps_cpu_mask = mask;
+        }
+
+        if (streq(rvalue, "disable")) {
+                cpu_set_reset(config->rps_cpu_mask);
+                return 0;
+        }
+
+        if (streq(rvalue, "all")) {
+                r = cpu_mask_add_all(config->rps_cpu_mask);
+                if (r < 0)
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "Failed to create CPU affinity mask representing \"all\" cpus, ignoring: %m");
+        } else
+                r = parse_cpu_set_extend(rvalue, config->rps_cpu_mask, /* warn= */ true, unit, filename, line, lvalue);
+
+        if (r < 0)
+                config->rps_cpu_mask = cpu_set_free(config->rps_cpu_mask);
```

Comment:
This clears previous assignment. For example,
```ini
ReceivePacketSteeringCPUMask=0-2
ReceivePacketSteeringCPUMask=foo
```
then the assignment `0-2` will be discarded.

So, we need to free the allocated one only when it is new.
How about something like the following?
```diff
diff --git a/src/shared/cpu-set-util.h b/src/shared/cpu-set-util.h
index eae82b11c9..618fe1b0a3 100644
--- a/src/shared/cpu-set-util.h
+++ b/src/shared/cpu-set-util.h
@@ -20,6 +20,8 @@ static inline void cpu_set_reset(CPUSet *a) {
 }
 
 CPUSet* cpu_set_free(CPUSet *c);
+DEFINE_TRIVIAL_CLEANUP_FUNC(CPUSet*, cpu_set_free);
+
 int cpu_set_add_all(CPUSet *a, const CPUSet *b);
 int cpu_set_add(CPUSet *a, unsigned cpu);
 
diff --git a/src/udev/net/link-config.c b/src/udev/net/link-config.c
index a197e56f69..8eee527bbf 100644
--- a/src/udev/net/link-config.c
+++ b/src/udev/net/link-config.c
@@ -1374,8 +1374,8 @@ int config_parse_rps_cpu_mask(
                 void *data,
                 void *userdata) {
 
-        LinkConfig *config = ASSERT_PTR(userdata);
-        CPUSet *mask = NULL;
+        _cleanup_(cpu_set_freep) CPUSet *allocated = NULL;
+        CPUSet *mask, **rps_cpu_mask = ASSERT_PTR(data);
         int r;
 
         assert(filename);
@@ -1383,33 +1383,40 @@ int config_parse_rps_cpu_mask(
         assert(rvalue);
 
         if (isempty(rvalue)) {
-                config->rps_cpu_mask = cpu_set_free(config->rps_cpu_mask);
+                *rps_cpu_mask = cpu_set_free(*rps_cpu_mask);
                 return 0;
         }
 
-        if (!config->rps_cpu_mask) {
-                mask = new0(CPUSet, 1);
-                if (!mask)
+        if (*rps_cpu_mask)
+                mask = *rps_cpu_mask;
+        else {
+                allocated = new0(CPUSet, 1);
+                if (!allocated)
                         return log_oom();
 
-                config->rps_cpu_mask = mask;
+                mask = allocated;
         }

         if (streq(rvalue, "disable")) {
-                cpu_set_reset(config->rps_cpu_mask);
+                cpu_set_reset(mask);
                 return 0;
         }
 
         if (streq(rvalue, "all")) {
-                r = cpu_mask_add_all(config->rps_cpu_mask);
-                if (r < 0)
+                r = cpu_mask_add_all(mask);
+                if (r < 0) {
                         log_syntax(unit, LOG_WARNING, filename, line, r,
                                    "Failed to create CPU affinity mask representing \"all\" cpus, ignoring: %m");
-        } else
-                r = parse_cpu_set_extend(rvalue, config->rps_cpu_mask, /* warn= */ true, unit, filename, line, lvalue);
+                        return 0;
+                }
+        } else {
+                r = parse_cpu_set_extend(rvalue, mask, /* warn= */ true, unit, filename, line, lvalue);
+                if (r < 0)
+                        return 0;
+        }
 
-        if (r < 0)
-                config->rps_cpu_mask = cpu_set_free(config->rps_cpu_mask);
+        if (allocated)
+                *rps_cpu_mask = TAKE_PTR(allocated);
 
         return 0;
 }
```

Reply:
Thanks :+1: